ddtrace/contrib/unittest/patch.py

```diff
@@ -0,0 +1,183 @@
+    _w(unittest, "TextTestResult.addSkip", add_skip_test_wrapper)
+
+    _w(unittest, "TestCase.run", start_test_wrapper_unittest)
+
+
+def unpatch():
+    if not getattr(unittest, "_datadog_patch", False):
+        return
+
+    _u(unittest.TextTestResult, "addSuccess")
+    _u(unittest.TextTestResult, "addFailure")
+    _u(unittest.TextTestResult, "addError")
+    _u(unittest.TextTestResult, "addSkip")
+    _u(unittest.TestCase, "run")
+
+    unittest._datadog_patch = False
+    _CIVisibility.disable()
+
+
+def _set_test_span_status(args, status):
+    test_item = args[0]
+    span = _extract_span(test_item)
+    if not span:
+        return
+    span.set_tag_str(test.STATUS, status)
+    if status == test.Status.FAIL.value:
+        exc_info = args[1]
+        span.set_exc_info(exc_info[0], exc_info[1], exc_info[2])
+    elif status == test.Status.SKIP.value:
+        span.set_tag_str(test.SKIP_REASON, _extract_test_skip_reason(args))
+
+
+def add_success_test_wrapper(func, instance, args, kwargs):
+    if is_unittest_support_enabled() and _is_valid_result(instance, args):
+        _set_test_span_status(args, test.Status.PASS.value)
+
+    return func(*args, **kwargs)
+
+
+def add_failure_test_wrapper(func, instance, args, kwargs):
+    if is_unittest_support_enabled() and _is_valid_result(instance, args):
+        _set_test_span_status(args, test.Status.FAIL.value)
+
+    return func(*args, **kwargs)
+
+
+def add_skip_test_wrapper(func, instance, args, kwargs):
+    result = func(*args, **kwargs)
+    if is_unittest_support_enabled() and _is_valid_result(instance, args):
+        _set_test_span_status(args, test.Status.SKIP.value)
+
+    return result
+
+
+def start_test_wrapper_unittest(func, instance, args, kwargs):
+    if is_unittest_support_enabled():
+        tracer = getattr(unittest, "_datadog_tracer", _CIVisibility._instance.tracer)
+        span = tracer._start_span(
+            ddtrace.config.unittest.operation_name,
+            service=_CIVisibility._instance._service,
+            resource="unittest.test",
+            span_type=SpanTypes.TEST,
+        )
+        span.set_tag_str(_EVENT_TYPE, SpanTypes.TEST)
+
+        span.set_tag_str(COMPONENT, COMPONENT_VALUE)
+        span.set_tag_str(SPAN_KIND, KIND)
+
+        span.set_tag_str(test.FRAMEWORK, FRAMEWORK)
+        span.set_tag_str(test.TYPE, SpanTypes.TEST)
+        suite_name = _extract_suite_name_from_test_method(instance)
+        span.set_tag_str(test.NAME, _extract_test_method_name(instance))
+        span.set_tag_str(test.SUITE, suite_name)
+        span.set_tag_str(test.MODULE, _extract_module_name_from_test_method(instance))
+
+        span.set_tag_str(test.STATUS, test.Status.FAIL.value)
+        span.set_tag_str(test.CLASS_HIERARCHY, suite_name)
+        _CIVisibility.set_codeowners_of(_extract_test_file_name(instance), span=span)
+
+        _store_span(instance, span)
+        result = func(*args, **kwargs)
```

Comment:
Can we use a contextmanager here? This will allow use to set the error status and message on a span if an exception is raised in `result = func(*args, **kwargs)`. 

ex: 

```
with tracer.trace(ddtrace.config.unittest.operation_name,
         service=_CIVisibility._instance._service,
         resource="unittest.test",
         span_type=SpanTypes.TEST) as span:
        
         span.set_tag_str(_EVENT_TYPE, SpanTypes.TEST)

        span.set_tag_str(COMPONENT, COMPONENT_VALUE)
        span.set_tag_str(SPAN_KIND, KIND)

        span.set_tag_str(test.FRAMEWORK, FRAMEWORK)
        span.set_tag_str(test.TYPE, SpanTypes.TEST)
        suite_name = _extract_suite_name_from_test_method(instance)
        span.set_tag_str(test.NAME, _extract_test_method_name(instance))
        span.set_tag_str(test.SUITE, suite_name)
        span.set_tag_str(test.MODULE, _extract_module_name_from_test_method(instance))

        span.set_tag_str(test.STATUS, test.Status.FAIL.value)
        span.set_tag_str(test.CLASS_HIERARCHY, suite_name)
        _CIVisibility.set_codeowners_of(_extract_test_file_name(instance), span=span)
        _store_span(instance, span)
        return result
   ```

Reply:
Isn't one of the points of `unittest`'s runner that it'll capture those exceptions (like `pytest`), to then mark the test as failed?

Using a context here would at least capture exceptions in `start_test_wrapper_unittest()` itself, even if the ones in `func()` end up getting caught, so maybe that's worth it.

Then again, I'm actually not sure what would happen at our intake if we had an exception here that still ended up sending a span. For example, if we had an exception setting the module, suite, or name tags, then imagine the intake would discard it anyway.

Reply:
Just switched to using a contextmanager instead, thanks!